dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -57,72 +58,124 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
-            // are routed inproc, otherwise with REGCLS_MULTIPLEUSE COM will launch a new process of the Win32 app for each invocation.
-            THROW_IF_FAILED(::CoRegisterClassObject(
-                AppModel::Identity::IsPackagedProcess() ? registeredClsid : storedComActivatorGuid,
-                winrt::make<AppNotificationManagerFactory>().get(),
-                CLSCTX_LOCAL_SERVER,
-                m_notificationHandlers ? REGCLS_MULTIPLEUSE : REGCLS_SINGLEUSE,
-                &m_notificationComActivatorRegistration));
+            hr = wil::ResultFromCaughtException();
+            throw;
         }
     }
 
     void AppNotificationManager::Unregister()
     {
-        auto lock{ m_lock.lock_exclusive() };
-        THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_notificationComActivatorRegistration, "Not Registered for App Notifications!");
-        m_notificationComActivatorRegistration.reset();
+        HRESULT hr{ S_OK };
+
+        auto logTelemetry{ wil::scope_exit([&]() {
+            std::wstring appId{};
+
+            try
+            {
+                appId = RetrieveNotificationAppId();
+            }
+            CATCH_LOG();
+
+            AppNotificationTelemetry::LogUnregister(hr, appId);
+        }) };
+
+        try
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_notificationComActivatorRegistration, "Not Registered for App Notifications!");
+            m_notificationComActivatorRegistration.reset();
+        }
+        catch (...)
+        {
+            hr = wil::ResultFromCaughtException();
+            throw;
+        }
     }
 
     void AppNotificationManager::UnregisterAll()
     {
-        Unregister();
+        HRESULT hr{ S_OK };
+        std::wstring appId{};
 
-        // Remove any Registrations from the Long Running Process that are necessary for Cloud toasts
-        if (!PushNotificationHelpers::IsPackagedAppScenario())
+        auto logTelemetry{ wil::scope_exit([&]() {
+            AppNotificationTelemetry::LogUnregisterAll(hr, appId);
+        }) };
+
+        try
         {
-            auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
-            THROW_IF_FAILED(notificationPlatform->RemoveToastRegistrationMapping(m_processName.c_str()));
-        }
+            Unregister();
+
+            // Remove any Registrations from the Long Running Process that are necessary for Cloud toasts
+            if (!PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                THROW_IF_FAILED(notificationPlatform->RemoveToastRegistrationMapping(m_processName.c_str()));
+            }
+
+            if (!AppModel::Identity::IsPackagedProcess())
+            {
+                appId = RetrieveNotificationAppId();
```

Comment:
Move this to outside: probably should be the first API call in the try block.

Reply:
appId is now initialized at construction now.